rfc/006-pex-info.md

```diff
@@ -0,0 +1,92 @@
+# RFC 006: Extend PEX messages with additional info
+
+## Changelog
+
+- 2021-07-07: Initial draft
+
+## Author
+
+- @xla
+
+## Context
+
+The majority of tendermint deployments run nodes with varying responsibilities
+offering subsets of all available channels/reactors. While this is accounted
+for in the p2p stack by not sending messages to a peer which did not advise the
+corresponding channel, it depends on the `NodeInfo` exchanged during
+a handshake after a connect. Consequently there currently isn't a protocol
+native way to do rudimentally discovery to find and connect to nodes offering
+certain capabilities.
+
+## Proposal
+
+This document proposes a change to the current set of PEX messages to support
+discovery based on advertised channels. In a way that is backward compabitible
+to avoid it requiring a hard-fork of any running networks.
+
+At the core is the introduction of a new request-response pair which besides
+the endpoint information (ip:port/url) carries the `NodeInfo` exchanged during
+handshake in the response. As that carries the channels active/supported on the
+peer a receiving node can apply filtering appropriate for local use-cases.
+Furthermore any extensions to the `NodeInfo` will be carried with the response
+enabling follow-up changes extending the ABCI to involve the application in the
+node discovery.
+
+### Message changes
+
+The recently introduced [v2 messages](https://github.com/tendermint/spec/pull/312)
+SHALL be replaced with versions which carry additional information about known
+peers.
+
+``` proto
+message PexNode {
+  // Encoded endpoint including protocol, id and path.
+  string   url  = 1 [(gogoproto.customname) = "URL"];
+  // Unaltered NodeInfo obtained during handshake with the peer.
+  NodeInfo info = 2;
```

Comment:
Whilst `NodeInfo` does provide helpful information in regards to `Channels` and `Network`, there is a lot of information that is quite irrelevant such as:
```
// ASCIIText fields
	Moniker string        `json:"moniker"` // arbitrary moniker
	Other   NodeInfoOther `json:"other"`   // other application specific data
```
also AFAIU, the `ListenAddr` and `NodeID` are already within the URL. 

In addition, there could be other information that could be worth considering s i.e. How many peers they are connected with and how many connections can they support. 

It might be better therefore to form a different struct to get passed rather than `NodeInfo`

Reply:
We've also had a debate internally as to whether channels should be able to dynamically change or whether they should remain static. They're currently static and not expected to change throughout the lifespan of the connection but making this something that is dynamic better supports node start up phases (where we may only run state sync or fast sync until we're ready to resume all services) as well as better handle conditions, for example, where a node begins to lag behind consensus and may temporarily close down services in reaction. This debate is something to also consider when thinking about which information to gossip amongst peers.

Reply:
@xla any thoughts on the above?

Reply:
In the above scenario, something like periodic checkins, like what some DHTs do, with peers to get an updated look at the network would be the best way to solve this. To a certain extent, we would be recreating a dht at this point because nodes may want a network wide outlook.

Reply:
> In addition, there could be other information that could be worth considering s i.e. How many peers they are connected with and how many connections can they support.

In my understanding on the role of `NodeInfo` these would be great candidates to extend `NodeInfo` with. I'd favour investing in this structure as it has a significant role in the protocol already and is guaranteed to be provided between two connected/handshaked peers. This makes it a strong vehicle to expand what is communicated and in turn what can be built on that information down the road.

> We've also had a debate internally as to whether channels should be able to dynamically change or whether they should remain static. They're currently static and not expected to change throughout the lifespan of the connection but making this something that is dynamic better supports node start up phases (where we may only run state sync or fast sync until we're ready to resume all services) as well as better handle conditions, for example, where a node begins to lag behind consensus and may temporarily close down services in reaction. This debate is something to also consider when thinking about which information to gossip amongst peers.

While in the back of my head as well, this is an orthogonal concern and should be designed in isolation and luckily doesn't negatively impact this document if left out. It's conceivable that an interactive protocol build in the future that communicates the latest active channel on a node to peers results in the local `NodeInfo` for that peer being updated. One thing is important no matter if the channels for a node are changing over the lifetime of a connection, when a PEX request-response interaction happen what comes through in the response is always gonna be a snapshot, therefore improvements like you hinting at can be plugged in later without the mechanism being different.

> @xla any thoughts on the above?

See paragraph above.

Reply:
> In the above scenario, something like periodic checkins, like what some DHTs do, with peers to get an updated look at the network would be the best way to solve this. To a certain extent, we would be recreating a dht at this point because nodes may want a network wide outlook.

Could you expand further what this DHT would key on?